plugins/outputs/opensearch/opensearch.go

```diff
@@ -0,0 +1,451 @@
+	}
+	if pipelineName != "" {
+		bulkIndexerConfig.Pipeline = pipelineName
+	}
+
+	return opensearchutil.NewBulkIndexer(bulkIndexerConfig)
+}
+
+func (o *Opensearch) manageTemplate(ctx context.Context) error {
+	tempReq := opensearchapi.CatTemplatesRequest{
+		Name: o.TemplateName,
+	}
+
+	resp, err := tempReq.Do(ctx, o.osClient.Transport)
+	if err != nil {
+		return fmt.Errorf("template check failed, template name: %s, error: %w", o.TemplateName, err)
+	}
+
+	templateExists := resp.Body != http.NoBody
+	templatePattern := o.IndexName
+
+	if strings.Contains(templatePattern, "%") {
+		templatePattern = templatePattern[0:strings.Index(templatePattern, "%")]
+	}
+
+	if strings.Contains(templatePattern, "{{") {
+		templatePattern = templatePattern[0:strings.Index(templatePattern, "{{")]
+	}
+
+	if templatePattern == "" {
+		return fmt.Errorf("template cannot be created for dynamic index names without an index prefix")
+	}
+
+	if (o.OverwriteTemplate) || (!templateExists) || (templatePattern != "") {
+		tp := templatePart{
+			TemplatePattern: templatePattern + "*",
+		}
+
+		t := template.Must(template.New("template").Parse(telegrafTemplate))
+		var tmpl bytes.Buffer
+
+		if err := t.Execute(&tmpl, tp); err != nil {
+			return err
+		}
+
+		indexTempReq := opensearchapi.IndicesPutTemplateRequest{
+			Name: o.TemplateName,
+			Body: strings.NewReader(tmpl.String()),
+		}
+		indexTempResp, errCreateTemplate := indexTempReq.Do(ctx, o.osClient.Transport)
+
+		if errCreateTemplate != nil || indexTempResp.StatusCode != 200 {
+			return fmt.Errorf("OpenSearch failed to create index template %s : %w", o.TemplateName, errCreateTemplate)
+		}
+
+		o.Log.Debugf("Template %s created or updated\n", o.TemplateName)
+	} else {
+		o.Log.Debug("Found existing OpenSearch template. Skipping template management")
+	}
+	return nil
+}
+
+func (o *Opensearch) GetReplacementKeys(indexName string, key string, replacement string) (string, []string) {
+	tagKeys := []string{}
+	startKey := "{{" + key
+	startTag := strings.Index(indexName, startKey)
+
+	for startTag >= 0 {
+		endTag := startTag + strings.Index(indexName[startTag:], "}}")
+
+		if endTag < 0 {
+			startTag = -1
+		} else {
+			tagName := indexName[startTag+len(startKey) : endTag]
+
+			var tagReplacer = strings.NewReplacer(
+				startKey+tagName+"}}", replacement,
+			)
+
+			indexName = tagReplacer.Replace(indexName)
+			tagKeys = append(tagKeys, strings.Trim(strings.TrimSpace(tagName), `"`))
+
+			startTag = strings.Index(indexName, startKey)
+		}
+	}
+
+	return indexName, tagKeys
+}
+
+func (o *Opensearch) GetIndexName(indexName string, eventTime time.Time, tagKeys []string, metricTags map[string]string) string {
+	tagValues := []interface{}{}
+
+	for _, key := range tagKeys {
+		if value, ok := metricTags[key]; ok {
+			tagValues = append(tagValues, value)
+		} else {
+			o.Log.Debugf("Tag '%s' not found, using '%s' on index name instead\n", key, o.DefaultTagValue)
```

Comment:
Please use the `%q` formatting directive to quote strings and remove all newlines from log messages!
```suggestion
			o.Log.Debugf("Tag %q not found, using %q on index name instead", key, o.DefaultTagValue)
```

Reply:
done